packages/@sanity/schema/src/sanity/extractSchema.ts

```diff
@@ -0,0 +1,669 @@
+      } satisfies ReferenceField,
+    })
+  })
+
+  return baseTypes
+}
+
+function createKeyField(): ObjectKeyValueField<OptionalField<StringField>> {
+  return {
+    type: 'objectKeyValue',
+    key: '_key',
+    value: optionalField({
+      type: 'string',
+    }),
+  }
+}
+
+function optionalField<T extends Field = Field>(field: T): OptionalField<T> {
+  return {
+    type: 'optional',
+    value: field,
+  }
+}
+function parseField(field: Exclude<FieldDefinition, SchemaField>): Field {
+  if (isObjectType(field)) {
+    const fields: ObjectKeyValueField[] = []
+    field.fields.forEach((f) => {
+      fields.push({
+        type: 'objectKeyValue',
+        key: f.name,
+        value: optionalField(parseField(f)),
+      })
+    })
+    fields.push({
+      type: 'objectKeyValue',
+      key: '_type',
+      value: {
+        type: 'string',
+        value: field.name,
+      },
+    } as ObjectKeyValueField<StringField>)
+
+    return {
+      type: field.type,
+      fields,
+    }
+  }
+
+  if (isArrayType(field)) {
+    return createArray(field)
+  }
+
+  if (isBlockType(field)) {
+    return createBlock(field)
+  }
+  if (isImageType(field)) {
+    return createImage(field)
+  }
+  if (isFileType(field)) {
+    return createFile(field)
+  }
+
+  if (isReferenceType(field)) {
+    return createReferenceFieldDefintion(field)
+  }
+
+  if (isCrossDatasetReferenceType(field)) {
+    return createCrossDatasetReferenceFieldDefintion(field)
+  }
+
+  if (typeMap[field.type]) {
+    return typeMap[field.type]
+  }
+
+  return {
+    type: 'reference',
+    to: field.type,
+  }
+}
+
+function isDocumentType(n: SchemaTypeDefinition): n is DocumentDefinition {
+  return n.type === 'document'
+}
+function isFieldDefinition(n: unknown): n is FieldDefinition {
+  return (
+    n !== null &&
+    typeof n === 'object' &&
+    'type' in n &&
+    (('fieldset' in n && typeof n.fieldset === 'string') ||
+      !('fieldset' in n) ||
+      typeof n.fieldset === 'undefined') &&
+    (('group' in n && (typeof n.group === 'string' || Array.isArray(n.group))) ||
+      !('group' in n) ||
+      typeof n.group === 'undefined')
+  )
+}
+
+function isObjectType(n: {type: string}): n is ObjectDefinition {
+  return n.type === 'object'
+}
+function isArrayType(n: {type: string}): n is ArrayDefinition {
+  return n.type === 'array'
+}
+function isBlockType(n: {type: string}): n is BlockDefinition {
+  return n.type === 'block'
+}
+function isReferenceType(n: {type: string}): n is ReferenceDefinition {
+  return n.type === 'reference'
+}
+function isCrossDatasetReferenceType(n: {type: string}): n is CrossDatasetReferenceDefinition {
+  return n.type === 'crossDatasetReference'
+}
+function isImageType(n: {type: string}): n is ImageDefinition {
+  return n.type === 'image'
+}
+function isFileType(n: {type: string}): n is FileDefinition {
+  return n.type === 'file'
+}
+
+type LiteralField = StringField | NumberField | BooleanField | NullField
```

Comment:
I think this should be referred to as a "primitive" instead of "literal" since in TS literal means the actual literal value, e.g. the number `1` or the string `"foo"`